routing/payment_store.go

```diff
@@ -0,0 +1,143 @@
+package routing
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lntypes"
+)
+
+var (
+	// paymentsBucketKey is the key used to access the top level bucket of
+	// the payment store. This bucket contains sub-buckets indexed by the
+	// payment hash of the active payments.
+	paymentsBucketKey = []byte("router-payment-store")
+
+	// paymentIDKey is used to access the paymentID of the current payment
+	// attempt sent to the switch. This key is only populated when the
+	// payment has status InFlight.
+	paymentIDKey = []byte("payment-id-key")
+
+	// paymentRouteKey is used to access the route of the current payment
+	// attempt sent to the switch. This key is only populated when the
+	// payment has status InFlight.
+	paymentRouteKey = []byte("payment-route-key")
+
+	// ErrNotFound is an error returned if the given payment hash was not
+	// found.
+	ErrNotFound = errors.New("payment hash not found")
+)
+
+// paymentStore is the persistent state machine used by the ChannelRouter to
+// track information about active payments. It is used to keep information
+// across restarts, needed to properly retry failed paymets, and handle payment
+// errors and successes encountered for a payment sent before the ChannelRouter
+// was restarted.
+type paymentStore struct {
```

Comment:
Should this be merged with the control tower?

Reply:
Since this depends on `routing.Route` this turned out to be quite tricky (getting rid of the dependency cycle made a huge diff). I prefer keeping it separate for now, as we also are going to add more to support payment resuming across restarts.

Reply:
By having the serialization code in the `channeldb` package, it wasn't that bad: https://github.com/lightningnetwork/lnd/compare/master...halseth:reliable-payments-router-state-machine-combined-control-tower-paymentstore?expand=1 

lmk if you prefer it that way!

Reply:
The later switch commits/PR remove the store all together (not sure it's safe to do this though), so as is, it gets removed anyway.

Reply:
`payment_store` removed, and this is now handled by `channeldb/control_tower`